### PR TITLE
improve bauhaus combo popup positioning at monitor edges

### LINKED ISSUE #11355: Cant select the All from the filter
**Describe the bug/issue**
Can't see the All when in darktable full screen view after selecting Rejected Only.

Workaround: select 2 star, then click again to select All. 

**To Reproduce**
With darktable in fullscreen (not lightable full screen). Select Rejected Only, then try to switch to All in the view selection.

**Expected behavior**
The selection populates in the visible screen.

**Screenshots**
image

**Platform**

* darktable version : 3.9.0+219~g06ad12bae
* OS                : Windows 11
* Memory            :  16Gb
* Graphics card     :  Nvidia 3060
* Graphics driver   :  511.79
* OpenCL installed  :  yes
* OpenCL activated  : ON

### PR DESCRIPTION
**Pull request by dterrahe**
fixes #11355

this fixes three things:
- if part of a bauhaus popup (for example combobox for filter/sort) is off the screen, then moving the mouse towards the screen edge will also show more of the popup.
- when you click&drag on a combo and then release on a different row, the row will get selected (also highlighting will work when dragging). so click-move-release instead of click-release-move-click-release
- double clicking a combo resets (this _sometimes_ worked previously)

Also, fix a memory leak by calling gtk_border_free on margin and padding and take the opportunity to merge combo_destroy and slider_destroy into one widget-class finalize. The problem could have been avoided by directly embedding margin and padding in dt_bauhaus_widget_t (instead of a pointer) but I guess that saves some memory for people who never show the side panels (and only costs two pointers and one level of indirection for everybody else).

### CODE REVIEW

**Comment by AlicVB** [comment]
@dterrahe : what was your idea here to fix #11355 ?
If I understand correctly the issue, the problem is that the popup top is higher than the main window, right ?
so, you wanted to ensure that the popup always stay in main window ?
But that will cause other problems (even without my recent work) as bauhaus popup is construct so it's initial position put the selected item at the exact position of the main widget one : the slider from the main widget exactly superposed the slider in the popup, same for combobox...

**Comment by dterrahe** [comment]
> bauhaus popup is construct so it's initial position put the selected item at the exact position of the main widget one

Yes, the code that you extended and replaced did exactly that. Except, of course, when called via a shortcut (and the widget maybe isn't visible or not even "realized" yet) in which case a shortcut caused the popup to show at the mouse cursor (and positioning is equally important so that the current selection is aligned with the cursor).

This PR starts with the popup in the same position as before, but then moves it down (or up) when the mouse nears the edge of the screen, if a part of the popup is not visible.

It has other fixed for combos as well; double click to reset and click (on the widget) and drag (in the popup) and then releasing immediately selects and closes.

BTW even if this doesn't cleanly merge against master, you can still check out this branch and see what it does.

**Comment by AlicVB** [comment]
thanks for your answer. I should have fixed the case where the widget is not realized here : https://github.com/darktable-org/darktable/pull/11380
Now the challenge will be to resolve conflict here... I'll give it a try, tomorrow if I can find some free time !

**Comment by dterrahe** [comment]
I'll look at conflicts.

**Comment by TurboGit** [comment]
And here are the conflicts :)

**Comment by AlicVB** [comment]
Great that you manage to rebase (and I promise to refrain to open bauhaus.c in the next days :))
Just to explain recent bauhaus changes, there has been multiple small issues to fix : 
- vertical alignment in cases where the widget exceed it's initial size
- possibility to set padding (useful in the case where the combobox has a background and a label)
- possibility to set margin (useful in the case where we have multiple different widgets in the same line, for consistency between widgets)

All that has done the recent bauhaus changes. Which has add some complexity in some part, but removed other complexity in other parts. For example all the hardcoded sizing/margins/padding management, with in many places a mix between multiplication and additions which makes the things works by the add of 1px here or there but breaks in many cases...

Of course this could have been done at the bauhaus level... with the cost that bauhaus widget would only be useful in iop/lib if they are the only widget in the line... That would have been a pity...

Now I'll try your PR and see if I can see anything to fix :)

**Comment by AlicVB** [comment]
ok, for the margins, it's my fault : a stupid mistake in css. Try to use something like : 
```
.bauhaus_combo_popup
{
  padding: 2px 0px 0px 2px;
}
```
Note the bottom padding that was missing ;) If it's ok for you, just integrate the change into your PR (or I can do another one at some point if you prefer)

File: src/bauhaus/bauhaus.c [hunk 1]
```diff
@@ -426,32 +384,54 @@ static gboolean dt_bauhaus_popup_scroll(GtkWidget *widget, GdkEventScroll *event
...
 
+  dt_bauhaus_widget_t *w = darktable.bauhaus->current;
   switch(w->type)
   {
     case DT_BAUHAUS_COMBOBOX:
+    {
+      GdkRectangle workarea;
+      gdk_monitor_get_workarea(gdk_display_get_monitor_at_window(gdk_window_get_display(window), window), &workarea);
+
+      gint dy = 0, move = darktable.bauhaus->mouse_y - ey;
```
Comment (AlicVB) [inline src/bauhaus/bauhaus.c:423]: use float here otherwise, you miss the small movements

File: src/bauhaus/bauhaus.c [hunk 2]
```diff
@@ -426,32 +384,54 @@ static gboolean dt_bauhaus_popup_scroll(GtkWidget *widget, GdkEventScroll *event
...
+      if(ey < darktable.bauhaus->mouse_y && wy < workarea.y)
+        dy = (workarea.y - wy) * move / (darktable.bauhaus->mouse_y + wy - workarea.y);
```
Comment (AlicVB) [inline src/bauhaus/bauhaus.c:425]: if I understand, you want that when your cursor reach the top of the main widow, the popup top reach the same point, right ?
In this case, that means that you want that a move of the cursor of `(darktable.bauhaus->mouse_y - workarea.y)` do a move of `(workarea.y - wy)` For a smaller movement of `move`, that would mean a popup movement of `dy = (workarea.y - wy) * move / (darktable.bauhaus->mouse_y - workarea.y);` (note the absence of the `wy` in the second part...

File: src/bauhaus/bauhaus.c [hunk 3]
```diff
@@ -426,32 +384,54 @@ static gboolean dt_bauhaus_popup_scroll(GtkWidget *widget, GdkEventScroll *event
...
+      if(ey > darktable.bauhaus->mouse_y && wy + allocation.height > workarea.y + workarea.height)
+        dy = (workarea.y + workarea.height - wy - allocation.height) * move / (darktable.bauhaus->mouse_y + wy - workarea.y - workarea.height);
+
       darktable.bauhaus->mouse_x = ex;
-      darktable.bauhaus->mouse_y = ey;
+      darktable.bauhaus->mouse_y = ey - dy;
```
Comment (AlicVB) [inline src/bauhaus/bauhaus.c:430]: `darktable.bauhaus->mouse_y` is the previous mouse position, so it should always be `ey`

**Review by AlicVB — COMMENTED** [review]
If I've understand well your goal, here is a proposal to fix the movements issues...
Note that I've done that only for "top" exceed, but same sort of thing should apply for bottom

In my test, that works, but I may very well have not understood your exact goal !

**Comment by dterrahe** [comment]
Thanks for trying to help.

ey (and therefore mouse_y) had padding->top deducted from it, so I just had to add it back in and it works perfectly again; as the mouse reaches the top of the monitor, the popup has _just_ reached it as well, if it needed to be shifted down.
```
      if(ey < darktable.bauhaus->mouse_y && wy < workarea.y)
        dy = (workarea.y - wy) * move / (darktable.bauhaus->mouse_y + wy + padding->top - workarea.y);
      if(ey > darktable.bauhaus->mouse_y && wy + allocation.height > workarea.y + workarea.height)
        dy = (workarea.y + workarea.height - wy - allocation.height) * move / (darktable.bauhaus->mouse_y + wy + padding->top - workarea.y - workarea.height);
```
The only thing I need to add; _workarea_ doesn't include the application bar at the bottom (for me, in kde) or presumably top of the screen (gnome), so it is possible if the mouse moves quickly above (or below) the "screen" we get a wrong way move. Or if pressing a shortcut while in that top or bottom area. Easy fix.

File: src/bauhaus/bauhaus.c (hunk 2 above)
Comment (dterrahe) [reply src/bauhaus/bauhaus.c:425]: I want to compare the previous mouse position on the _screen_ to the edge. mouse_y is relative to the popup, which at this point is still a little off the screen (and therefore the mouse can't reach it).

File: src/bauhaus/bauhaus.c (hunk 3 above)
Comment (dterrahe) [reply src/bauhaus/bauhaus.c:430]: mouse_y is relative to the _window_. But we are _moving_ the window underneath the mouse. So the physical mouse is now in a slightly different position relative to the window.

**Comment by AlicVB** [comment]
yep, that seems to work (although I will have to do some little drawings to understand why ;))
but here, you still need to put dy/move in float, to get smooth movement

**Comment by dterrahe** [comment]
> you still need to put dy/move in float, to get smooth movement

you are right; not for the small mouse moves, but to force the calculation to use float. :-)

**Comment by dterrahe** [comment]
Done.

The popup window still seems too small for combos (possibly needs further correcting for padding/margins?). Unless that's a regression in this PR. @AlicVB could you have a look please?

Also, sometimes the highlighted item seems to be the one just below the mouse (and sometimes a different one gets selected when releasing mouse). If this is not a problem in master, please let me know and I'll find where I broke it...

**Comment by dterrahe** [comment]
> .bauhaus_combo_popup
> {
>   padding: 2px 0px 0px 2px;
> }

Sorry, I missed this suggestion earlier, but it seems this is already in master (so doesn't solve the popup sizing issue)?

**Comment by AlicVB** [comment]
For the bottom padding of combobox popup, as said before in this thread, it's an stupid mistake from my part in css, where I forgot to set some bottom padding. Correct code should be something like : (note the add of a bottom margin)
```
.bauhaus_combo_popup
{
  padding: 2px 0px 2px 2px;
}
```
@Nilvus is aware of that and plan to fix it in a PR soon

For the offset, it's my fault too ! In case where we have a base widget that in higher than the text height + padding + margin, we need to center the text, so we need basically to add some top padding. This padding is also added to the popup in order to ensure that the popup top match with main widget top for first element... and I've forgotten to add that to hovered and select...
to fix that point, line 1801 should be `? ((darktable.bauhaus->end_mouse_y - w->top_gap) / darktable.bauhaus->line_height)`
and line 1991 : `const int hovered = (darktable.bauhaus->mouse_y - w->top_gap) / darktable.bauhaus->line_height;`
If it's ok for you just integrate that to your last commit (or I can do it later if you prefer) !

**Comment by dterrahe** [comment]
still feels a little cramped to me; the j is touching the border
image

and the point of the mouse cursor is higher than the bottom of the p, but still the line below selected
image

**Comment by TurboGit** [comment]
> and the point of the mouse cursor is higher than the bottom of the p, but still the line below selected

Yes, I have seen that too, some pixels offset but the computation seems OK, so I'm not sure how to fix that at this point.

**Comment by AlicVB** [comment]
OK, I think I've found the reason of this small offset... but no solution to propose atm :(
The root cause is that `pango_cairo_show_layout` used to draw text in bauhaus draw the text a little bit too low. In fact for a text like "lljj", here, it return a text height of 15px. Althought the draw text take 13px. And the problem is that the margins are 2px on top and 0 on bottom...

To show that, you can add something like : 
```
cairo_save(cr);
        cairo_set_source_rgb(cr, 1, 0, 0);
        cairo_rectangle(cr, j*4 + 1, ht * k + w->top_gap, 4, ht);
        cairo_fill(cr);
        cairo_restore(cr);
```
at line 1998 (inside the `for` loop used to draw each text line)
Note that same thing can be done with same result for main combobox... and as the text in those combobox are exactly aligned with "classic" label, I would say that gtklabel use the same textdrawing solution...

**Comment by AlicVB** [comment]
ok, I've also missed to add the "gap" to the popup height computation...
We need to add `tmp.height += w->top_gap;` line 2372. Note that I've checked, and that also ok if the combobox is not realized (the gap is 0)

Can I let you add that too ? (sorry to let you fix my mistake...)

For the offset, we have spoken earlier, that seems to be a design choice of cairo/pango to include a "minimal" line spacing at the top. The problem is that I've not found a reliable way to determine that spacing value...
So I would say that we can let it like that and just let @Nilvus add a milttle bigger bottom padding (anyway he wanted to finetune that point once this merged)

**Review by TurboGit — APPROVED** [review]
Looks good, thanks! Let's move forward and let @Nilvus finalize tthe CSS part.

**Pull request merged**